### ISSUE TITLE
Building CheriBSD with -O2 for CheriABI leads to linkage errors

### ISSUE THREAD

**Issue by rwatson** [CTSRD-CHERI/llvm#151]
When building CheriBSD pure-capability binaries with ```-O2```:
```
--- a/share/mk/bsd.cheri.mk
+++ b/share/mk/bsd.cheri.mk
@@ -31,7 +31,7 @@ OBJCOPY:=     elfcopy
 _CHERI_CC+=    -mabi=sandbox -mxgot
 LIBDIR:=       /usr/libcheri
 ROOTOBJDIR=    ${.OBJDIR:S,${.CURDIR},,}${SRCTOP}/worldcheri${SRCTOP}
-CFLAGS+=       -O0
+CFLAGS+=       -O2
 .if ${MK_CHERI_LINKER} == "yes"
 _CHERI_CC+=    -cheri-linker
 CFLAGS+=       -Wno-error
```

I get the following error during the link phase:
```
/home/rnw24/sdk/sdk/bin/clang -g -integrated-as --target=cheri-unknown-freebsd -msoft-float --sysroot=/home/rnw24/obj/mips.mips64/home/rnw24/git/cheribsd/tmp -mabi=sandbox -mxgot -cheri-linker -O -pipe -G0 -DINET6 -DHAVE_NET_PFVAR_H -DHAVE_CONFIG_H -DTCPDUMP_PRINT_ONLY -DTCPDUMP_HELPER -DCHERI_TCPDUMP_INTERNAL -D_U_="__attribute__((unused))" -I/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump -I/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../lib/libcheri -I/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../usr.sbin/tcpdump/tcpdump -I/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../usr.sbin/tcpdump/cheri_tcpdump -mxgot -DLBL_ALIGN -O2 -Wno-error -g -std=gnu99 -Wsystem-headers -Werror -Wall -Wno-format-y2k -W -Wno-unused-parameter -Wstrict-prototypes -Wmissing-prototypes -Wpointer-arith -Wno-uninitialized -Wno-pointer-sign -Wformat=2 -Wno-format-extra-args -Werror -Wno-empty-body -Wno-string-plus-int -Wno-unused-const-variable -Wno-tautological-compare -Wno-unused-value -Wno-parentheses-equality -Wno-unused-function -Wno-enum-conversion -Wno-unused-local-typedef -Wno-format-nonliteral -Qunused-arguments -Wno-cast-qual  -Wl,--no-warn-mismatch -Wl,--script=/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../lib/libc_cheri/sandbox.ld  -nostdlib -static -o tcpdump-helper.full tcpdump-helper.o pcap_datalink_val_to_name.o pcap-eproto_db.o stub.o util.o addrtoname.o af.o checksum.o cpack.o gmpls.o in_cksum.o ipproto.o l2vpn.o nlpid.o oui.o parsenfsfh.o print.o print-802_11.o print-802_15_4.o print-ah.o print-ahcp.o print-aodv.o print-aoe.o print-ap1394.o print-arcnet.o print-arp.o print-ascii.o print-atalk.o print-atm.o print-beep.o print-bfd.o print-bgp.o print-bootp.o print-bt.o print-calm-fast.o print-carp.o print-cdp.o print-cfm.o print-chdlc.o print-cip.o print-cnfp.o print-dccp.o print-decnet.o print-domain.o print-dtp.o print-dvmrp.o print-eap.o print-egp.o print-eigrp.o print-enc.o print-esp.o print-ether.o print-fddi.o print-forces.o print-fr.o print-ftp.o print-geneve.o print-geonet.o print-gre.o print-hsrp.o print-http.o print-icmp.o print-igmp.o print-igrp.o print-ip.o print-ip6.o print-ipcomp.o print-ipfc.o print-ipnet.o print-ipx.o print-isakmp.o print-isoclns.o print-juniper.o print-krb.o print-l2tp.o print-lane.o print-ldp.o print-llc.o print-lldp.o print-lmp.o print-loopback.o print-lspping.o print-lwapp.o print-lwres.o print-m3ua.o print-mobile.o print-mpcp.o print-mpls.o print-mptcp.o print-msdp.o print-msnlb.o print-nfs.o print-ntp.o print-null.o print-olsr.o print-openflow.o print-openflow-1.0.o print-ospf.o print-otv.o print-pgm.o print-pim.o print-pktap.o print-ppi.o print-ppp.o print-pppoe.o print-pptp.o print-radius.o print-raw.o print-rip.o print-rpki-rtr.o print-rrcp.o print-rsvp.o print-rtsp.o print-rx.o print-sctp.o print-sflow.o print-sip.o print-sl.o print-sll.o print-slow.o print-smb.o print-smtp.o print-snmp.o print-stp.o print-sunatm.o print-sunrpc.o print-symantec.o print-syslog.o print-tcp.o print-telnet.o print-tftp.o print-timed.o print-tipc.o print-token.o print-udld.o print-udp.o print-vjc.o print-vqp.o print-vrrp.o print-vtp.o print-vxlan.o print-wb.o print-zephyr.o print-zeromq.o signature.o smbutil.o util-print.o print-babel.o print-dhcp6.o print-frag6.o print-icmp6.o print-ip6opts.o print-mobility.o print-ospf6.o print-ripng.o print-rt6.o print-pflog.o print-pfsync.o   -lc_cheri
print-isakmp.o: In function `ikev2_print':
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/print-isakmp.c:2868: relocation truncated to fit: R_MIPS_GOT_PAGE against `ikev2_sub_print'
print-isakmp.o: In function `ikev1_t_print':
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/print-isakmp.c:1271: relocation truncated to fit: R_MIPS_GOT_PAGE against `ikev1_attrmap_print'
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/print-isakmp.c:1274: relocation truncated to fit: R_MIPS_GOT_PAGE against `ikev1_attr_print'
print-isakmp.o: In function `ikev1_nonce_print':
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/print-isakmp.c:1616: relocation truncated to fit: R_MIPS_GOT_PAGE against `ike_show_somedata'
print-isakmp.o: In function `ikev1_n_print':
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/print-isakmp.c:1737: relocation truncated to fit: R_MIPS_GOT_PAGE against `ikev1_attrmap_print'
print-isakmp.o: In function `ikev2_sa_print':
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/print-isakmp.c:2001: relocation truncated to fit: R_MIPS_GOT_PAGE against `ikev2_sub_print'
print-isakmp.o: In function `ikev2_cert_print':
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/print-isakmp.c:2129: relocation truncated to fit: R_MIPS_GOT_PAGE against `ikev2_gen_print'
print-isakmp.o: In function `ikev2_cr_print':
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/print-isakmp.c:2139: relocation truncated to fit: R_MIPS_GOT_PAGE against `ikev2_gen_print'
print-isakmp.o: In function `ikev2_auth_print':
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/print-isakmp.c:2169: relocation truncated to fit: R_MIPS_GOT_PAGE against `ike_show_somedata'
print-isakmp.o: In function `ikev2_nonce_print':
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/print-isakmp.c:2198: relocation truncated to fit: R_MIPS_GOT_PAGE against `ike_show_somedata'
print-isakmp.o: In function `ikev2_n_print':
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/print-isakmp.c:2406: additional relocation overflows omitted from the output
clang-3.8: error: linker command failed with exit code 1 (use -v to see invocation)
*** Error code 1

Stop.
make[4]: stopped in /home/rnw24/git/cheribsd/libexec/tcpdump-helper
```

At ```-O0```, this problem does not arise.

**Comment by rwatson** [comment]
This also occurs at ```-O1```.

**Comment by davidchisnall** [comment]
This is usually caused by  an issue with out linker not handling relaxations for symbols that are too far away.  Adding `-mxgot` should fix it.

**Comment by rwatson** [comment]
Hmm. It looks like both the original ```.o``` command line and linking command line includes ```-mxgot```:
```
/home/rnw24/sdk/sdk/bin/clang -g -integrated-as --target=cheri-unknown-freebsd  
-msoft-float --sysroot=/home/rnw24/obj/mips.mips64/home/rnw24/git/cheribsd/tmp -
mabi=sandbox -mxgot -cheri-linker -O -pipe -G0  -DINET6 -DHAVE_NET_PFVAR_H -DHAV
E_CONFIG_H -DTCPDUMP_PRINT_ONLY -DTCPDUMP_HELPER -DCHERI_TCPDUMP_INTERNAL -D_U_=
"__attribute__((unused))" -I/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../.
./contrib/tcpdump -I/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../lib/li
bcheri -I/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../usr.sbin/tcpdump/
tcpdump -I/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../usr.sbin/tcpdump
/cheri_tcpdump -mxgot -DLBL_ALIGN -O1 -Wno-error  -g -std=gnu99 -Wsystem-headers
 -Werror -Wall -Wno-format-y2k -W -Wno-unused-parameter -Wstrict-prototypes -Wmi
ssing-prototypes -Wpointer-arith -Wno-uninitialized -Wno-pointer-sign -Wformat=2
 -Wno-format-extra-args -Werror -Wno-empty-body -Wno-string-plus-int -Wno-unused
-const-variable -Wno-tautological-compare -Wno-unused-value -Wno-parentheses-equ
ality -Wno-unused-function -Wno-enum-conversion -Wno-unused-local-typedef -Wno-f
ormat-nonliteral  -Qunused-arguments -Wno-cast-qual  -c /home/rnw24/git/cheribsd
/libexec/tcpdump-helper/../../contrib/tcpdump/addrtoname.c -o addrtoname.o
...
/home/rnw24/sdk/sdk/bin/clang -g -integrated-as --target=cheri-unknown-freebsd -
msoft-float --sysroot=/home/rnw24/obj/mips.mips64/home/rnw24/git/cheribsd/tmp -m
abi=sandbox -mxgot -cheri-linker -O -pipe -G0 -DINET6 -DHAVE_NET_PFVAR_H -DHAVE_
CONFIG_H -DTCPDUMP_PRINT_ONLY -DTCPDUMP_HELPER -DCHERI_TCPDUMP_INTERNAL -D_U_="_
_attribute__((unused))" -I/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../
contrib/tcpdump -I/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../lib/libc
heri -I/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../usr.sbin/tcpdump/tc
pdump -I/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../usr.sbin/tcpdump/c
heri_tcpdump -mxgot -DLBL_ALIGN -O1 -Wno-error -g -std=gnu99 -Wsystem-headers -W
error -Wall -Wno-format-y2k -W -Wno-unused-parameter -Wstrict-prototypes -Wmissi
ng-prototypes -Wpointer-arith -Wno-uninitialized -Wno-pointer-sign -Wformat=2 -W
no-format-extra-args -Werror -Wno-empty-body -Wno-string-plus-int -Wno-unused-co
nst-variable -Wno-tautological-compare -Wno-unused-value -Wno-parentheses-equali
ty -Wno-unused-function -Wno-enum-conversion -Wno-unused-local-typedef -Wno-form
at-nonliteral -Qunused-arguments -Wno-cast-qual  -Wl,--no-warn-mismatch -Wl,--sc
ript=/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../lib/libc_cheri/sandbo
x.ld  -nostdlib -static -o tcpdump-helper.full tcpdump-helper.o pcap_datalink_va
l_to_name.o pcap-eproto_db.o stub.o util.o addrtoname.o af.o checksum.o cpack.o
gmpls.o in_cksum.o ipproto.o l2vpn.o nlpid.o oui.o parsenfsfh.o print.o print-80
2_11.o print-802_15_4.o print-ah.o print-ahcp.o print-aodv.o print-aoe.o print-a
p1394.o print-arcnet.o print-arp.o print-ascii.o print-atalk.o print-atm.o print
-beep.o print-bfd.o print-bgp.o print-bootp.o print-bt.o print-calm-fast.o print
-carp.o print-cdp.o print-cfm.o print-chdlc.o print-cip.o print-cnfp.o print-dcc
p.o print-decnet.o print-domain.o print-dtp.o print-dvmrp.o print-eap.o print-eg
p.o print-eigrp.o print-enc.o print-esp.o print-ether.o print-fddi.o print-force
s.o print-fr.o print-ftp.o print-geneve.o print-geonet.o print-gre.o print-hsrp.
o print-http.o print-icmp.o print-igmp.o print-igrp.o print-ip.o print-ip6.o pri
nt-ipcomp.o print-ipfc.o print-ipnet.o print-ipx.o print-isakmp.o print-isoclns.
o print-juniper.o print-krb.o print-l2tp.o print-lane.o print-ldp.o print-llc.o 
print-lldp.o print-lmp.o print-loopback.o print-lspping.o print-lwapp.o print-lw
res.o print-m3ua.o print-mobile.o print-mpcp.o print-mpls.o print-mptcp.o print-
msdp.o print-msnlb.o print-nfs.o print-ntp.o print-null.o print-olsr.o print-ope
nflow.o print-openflow-1.0.o print-ospf.o print-otv.o print-pgm.o print-pim.o pr
int-pktap.o print-ppi.o print-ppp.o print-pppoe.o print-pptp.o print-radius.o pr
int-raw.o print-rip.o print-rpki-rtr.o print-rrcp.o print-rsvp.o print-rtsp.o pr
int-rx.o print-sctp.o print-sflow.o print-sip.o print-sl.o print-sll.o print-slo
w.o print-smb.o print-smtp.o print-snmp.o print-stp.o print-sunatm.o print-sunrp
c.o print-symantec.o print-syslog.o print-tcp.o print-telnet.o print-tftp.o prin
t-timed.o print-tipc.o print-token.o print-udld.o print-udp.o print-vjc.o print-
vqp.o print-vrrp.o print-vtp.o print-vxlan.o print-wb.o print-zephyr.o print-zer
omq.o signature.o smbutil.o util-print.o print-babel.o print-dhcp6.o print-frag6
.o print-icmp6.o print-ip6opts.o print-mobility.o print-ospf6.o print-ripng.o pr
int-rt6.o print-pflog.o print-pfsync.o   -lc_cheri
addrtoname.o: In function `etheraddr_string':
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/addrtoname
.c:499: relocation truncated to fit: R_MIPS_GOT_PAGE against `lookup_emem'
addrtoname.o: In function `le64addr_string':
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/addrtoname
.c:546: relocation truncated to fit: R_MIPS_GOT_PAGE against `lookup_bytestring'
addrtoname.o: In function `linkaddr_string':
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/addrtoname
.c:581: relocation truncated to fit: R_MIPS_GOT_PAGE against `lookup_bytestring'
addrtoname.o: In function `protoid_string':
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/addrtoname
.c:633: relocation truncated to fit: R_MIPS_GOT_PAGE against `lookup_protoid'   
addrtoname.o: In function `isonsap_string':
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/addrtoname
.c:663: relocation truncated to fit: R_MIPS_GOT_PAGE against `lookup_nsap'
addrtoname.o: In function `init_addrtoname':
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/addrtoname
.c:1173: relocation truncated to fit: R_MIPS_GOT_PAGE against `init_etherarray' 
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/addrtoname
.c:1174: relocation truncated to fit: R_MIPS_GOT_PAGE against `init_servarray'  
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/addrtoname
.c:1175: relocation truncated to fit: R_MIPS_GOT_PAGE against `init_eprotoarray'
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/addrtoname
.c:1176: relocation truncated to fit: R_MIPS_GOT_PAGE against `init_protoidarray
'
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/addrtoname
.c:1177: relocation truncated to fit: R_MIPS_GOT_PAGE against `init_ipxsaparray'
addrtoname.o: In function `init_etherarray':
/home/rnw24/git/cheribsd/libexec/tcpdump-helper/../../contrib/tcpdump/addrtoname
.c:897: additional relocation overflows omitted from the output
clang-3.8: error: linker command failed with exit code 1 (use -v to see invocati
on)
*** Error code 1
```

Preprocessed source code: https://gist.github.com/rwatson/7fd7d7b8f390ca69e199

Tagging @brooksdavis as he will also be interested in the eventual outcome of this issue.

**Comment by davidchisnall** [comment]
I believe that this was fixed in 558040666503ceba9111258edfe2efa2de6492bf
